### PR TITLE
Upload update on candidate labeler

### PR DESCRIPTION
**Pull request by danich1**
This PR contains code to automatically label disease gene candidates. This is a follow up on my old [snorkeling](https://github.com/greenelab/snorkeling) project where the goal here is to be able to auto-label large amount of data quickly. Then after labeling I am able to train deep learning models (future PR to come). Main review here is the notebook, but feel free to look at the other functions I have.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Snorkeling-Full-Text
 
-This is an upgraded version of the original [snorkeling project](https://github.com/greenelab/snorkeling). 
+This is an upgraded version of the original [snorkeling project](https://github.com/greenelab/snorkeling).
```
Comment (ajlee21) [inline README.md:3]: Sorry its been a while, so here you're using full text vs previously using abstracts to learn relationships (like disease gene relationships). Where you I recall you have an input X matrix that is abstract x label function and the Y = 1/0 there is a relationship

Comment (ajlee21) [reply README.md:3]: Also super nit-picky. Just noticed a few typos in your old repo readme if you feel like updating it

Comment (ajlee21) [reply README.md:3]: Whats the reason this repo is being deprecated? Is full text just much better?

File: label_candidates/label_candidates.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+where sentence_id not in ({candidates}) and
+section = 'title' or section ='abstract'
+"""
+
+candidate = pd.read_sql(sql, database_str)
+print(candidate.shape)
+candidate.head()
+
+# +
+sen_id_map = ",".join(map(str, candidate.sentence_id.tolist()))
```
Comment (ajlee21) [inline label_candidates/label_candidates.py:60]: Nice clean code. realize much of this is redundant to the old repo, but if people are going to be looking at this one in the future, might be helpful to add more comments about what this notebook is doing..Or maybe you plan to add this in the readme later

File: snorkeling_helper/candidates_helper/encode_candidates_modules.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def encode_lemmas(
+    candidate_df,
+    stopwords,
+    word_mapper,
+    entity_fieldnames=["disease_start", "disease_end", "gene_start", "gene_end"],
+    entity_one="DISEASE_ENTITY",
+    entity_two="GENE_ENTITY",
+):
```
Comment (ajlee21) [inline snorkeling_helper/candidates_helper/encode_candidates_modules.py:11]: Perhaps adding a description here would be helpful. I forget what these lemmas are

File: snorkeling_helper/candidates_helper/label_candidates_modules.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+def char_to_word(row):
```
Comment (ajlee21) [inline snorkeling_helper/candidates_helper/label_candidates_modules.py:1]: Looks like you also have this defined in your notebook. Maybe delete the one in the notebook?

File: snorkeling_helper/label_functions/disease_gene/disease_associates_gene.py [hunk 5]
```diff
@@ -0,0 +1,1164 @@
...
+for row in pair_df.itertuples():
+    if not row.sources or pd.isnull(row.sources):
+        continue
+    for source in row.sources.split("|"):
+        key = str(row.entrez_gene_id), row.doid_id, source
+        knowledge_base.add(key)
+
+
+@labeling_function()
+def LF_HETNET_DISEASES(c):
```
Comment (ajlee21) [inline snorkeling_helper/label_functions/disease_gene/disease_associates_gene.py:37]: Are these label functions the same as previously?

**Review by ajlee21 — APPROVED** [review]
All the code looks good to me, just had more conceptual questions

**Comment by ajlee21** [comment]
Not sure if this will be helpful, but looks like they're also using labeling functions but to create new labeled training data: https://github.com/megagonlabs/ruler

File: README.md [hunk 6]
```diff
@@ -11,23 +11,23 @@ First run `git clone https://github.com/greenelab/pubtator` in the same director
...
-Once everything has been installed, type following command in the terminal: 
+
+Once everything has been installed, type following command in the terminal:
 
 ```bash
 bash install.sh
-``` 
-_Note_: 
+```
+_Note_:
```
Comment (ben-heil) [inline README.md:22]: You may want to link Windows users to Windows Bash here

File: label_candidates/label_candidates.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            "pos_tags": "pos_tag",
+        },
+    )
+    .assign(
+        lemma=lambda x: x.lemma.apply(lambda s: s.split("|")),
+        word=lambda x: x.word.apply(lambda s: s.split("|")),
+        pos_tag=lambda x: x.pos_tag.apply(lambda s: s.split("|")),
+    )
+)
+fixed_annotated_df
```
Comment (ben-heil) [inline label_candidates/label_candidates.py:228]: Is it okay to be getting NaN sentence IDs here?

File: snorkeling_helper/candidates_helper/encode_candidates_modules.py (hunk 3 above)
Comment (ben-heil) [reply snorkeling_helper/candidates_helper/encode_candidates_modules.py:11]: +1, except that I know what a lemma is but not why you would encode one

File: snorkeling_helper/label_functions/disease_gene/disease_associates_gene.py [hunk 8]
```diff
@@ -0,0 +1,1164 @@
...
+    "recombinant human",
+    "novel {{B}} gene mutation",
+    "pleotropic effects on",
+}
+
+context_change_keywords = {
+    ", but",
+    ", whereas",
+    "; however,",
+}
```
Comment (ben-heil) [inline snorkeling_helper/label_functions/disease_gene/disease_associates_gene.py:489]: Oh wow, 300 lines of indicator words must have taken awhile

**Review by ben-heil — APPROVED** [review]
Looks good!

File: README.md (hunk 1 above)
Comment (danich1) [reply README.md:3]: >Sorry its been a while, so here you're using full text vs previously using abstracts to learn relationships (like disease gene relationships). Where you I recall you have an input X matrix that is abstract x label function and the Y = 1/0 there is a relationship

Yes for this section its individual sentences from abstracts X label function input. Once I have this matrix I'm feeding it into the generative model to estimate "accurate" training labels I can use for downstream machine learning (a few PRs from now).

>Also super nit-picky. Just noticed a few typos in your old repo readme if you feel like updating it

Feel free to suggest changes or pass them on to me. Appreciate the proof reading.

>Whats the reason this repo is being deprecated? Is full text just much better? 

That's what this new angle is going to find out. Given the programming bugs in my last repository, I decided that just starting fresh was a better option. I had to go all the way back to square one, so depreciating the old repo just made sense in my head.

File: label_candidates/label_candidates.py (hunk 2 above)
Comment (danich1) [reply label_candidates/label_candidates.py:60]: I agree it would be easier to add description into this notebook, so I'll add that. Good catch!

File: snorkeling_helper/candidates_helper/encode_candidates_modules.py (hunk 3 above)
Comment (danich1) [reply snorkeling_helper/candidates_helper/encode_candidates_modules.py:11]: >Perhaps adding a description here would be helpful. I forget what these lemmas are

Lemmas are the root form of words i.e. The word "are" -> "be" or "walked" -> "walk". In NLP taking the root form of worlds helps lower the dimensionality of my training set when I get ready to train the network as well as keep sentences simple.

>+1, except that I know what a lemma is but not why you would encode one

Ah this step is for the transformer. In order to get the network to understand my sentences I needed to have an encoding layer to get a training matrix of (batch size, sequence size, embedding size). Right now it doesn't make much sense but one-two PRs from now it will click.

File: snorkeling_helper/label_functions/disease_gene/disease_associates_gene.py (hunk 5 above)
Comment (danich1) [reply snorkeling_helper/label_functions/disease_gene/disease_associates_gene.py:37]: >Are these label functions the same as previously?

Yes these notebooks are the same as the previous notebook; however, one catch here is that they have been debugged and updated to snorkel's most current version. I just wanted to separate the previous repository with this new repository, so people have a new start if they want to begin from scratch.

File: README.md (hunk 1 above)
Comment (ajlee21) [reply README.md:3]: Just curious, I understand the logic for starting fresh, but is there a reason that you're not using abstracts here again..or was that part of the previous issue?

Comment (danich1) [reply README.md:3]: >Just curious, I understand the logic for starting fresh, but is there a reason that you're not using abstracts here again..or was that part of the previous issue?

No I am using abstracts as well. Its just that pubtator updates every month or so. So my previous repo is outdated by a few years.

File: label_candidates/label_candidates.py (hunk 7 above)
Comment (danich1) [reply label_candidates/label_candidates.py:228]: Yeah its fine. As those NaNs are the sentences that failed to be processed via my pipeline. Processing full text and abstracts is not a clean process.
